File: purecache/example/bad/bad.go
// Package bad collects functions that claim purity but are not pure. Running
// `purecache example/bad` should reject every one of them.
package bad

import (
	"math/rand"
	"os"
	"time"
)

var callCount int

type Buffer struct{ Data []byte }

//cache:pure
func Counter(n int) int {
	callCount++ // reads and writes a package-level var
	return n + callCount
}

//cache:pure
func Stamped(n int) int {
	return n + int(time.Now().Unix()) // reaches outside its arguments
}

//cache:pure
func Noisy(n int) int {
	return n + rand.Intn(10) // not a function of its arguments
}

//cache:pure
func Env(key string) string {
	return os.Getenv(key) // reads ambient state
}

//cache:pure
func Apply(n int, f func(int) int) int {
	return f(n) // a func value has no hashable content
}

//cache:pure
//cache:deep b
func Clobber(b *Buffer) int {
	b.Data = append(b.Data, 1) // mutates through a parameter
	return len(b.Data)
}

//cache:pure
func Unmarked(xs []int) int {
	return len(xs) // reference-typed parameter with no //cache:deep promise
}

//cache:pure
func Concurrent(n int) int {
	ch := make(chan int, 1)
	go func() { ch <- n * 2 }()
	return <-ch
}

File: zigcache/example/bad.zig
//! Functions that claim purity but are not. zigcache must reject every one.

const std = @import("std");

var call_count: u64 = 0;

const Buffer = struct { data: []u8 };

///cache:pure
pub fn counter(n: u64) u64 {
    call_count += 1;
    return n + call_count;
}

///cache:pure
pub fn stamped(n: i64) i64 {
    return n + std.time.timestamp();
}

///cache:pure
pub fn noisy(n: u64) u64 {
    return n + std.crypto.random.int(u8);
}

///cache:pure
pub fn logged(n: u64) u64 {
    std.debug.print("{d}\n", .{n});
    return n;
}

///cache:pure
pub fn applied(n: u64, f: *const fn (u64) u64) u64 {
    return f(n);
}

///cache:pure
///cache:deep b
pub fn clobber(b: *Buffer) usize {
    b.data = b.data[0..0];
    return b.data.len;
}

///cache:pure
pub fn unmarked(xs: []const u64) usize {
    return xs.len;
}

///cache:pure
pub fn generic(x: anytype) usize {
    return @sizeOf(@TypeOf(x));
}

///cache:pure
pub fn peeked(addr: usize) u8 {
    const p: *const u8 = @ptrFromInt(addr);
    return p.*;
}

///cache:pure
pub fn sneaky(n: u64) u64 {
    return n + helper();
}

fn helper() u64 {
    return @intFromPtr(&call_count);
}
